src/plugins/intel_gpu/src/graph/non_max_suppression.cpp

```diff
@@ -81,4 +85,71 @@ std::string non_max_suppression_inst::to_string(non_max_suppression_node const&
     return description.str();
 }
 
+// -----------------------------------------------
+// non_max_suppression_gather
+// -----------------------------------------------
+GPU_DEFINE_PRIMITIVE_TYPE_ID(non_max_suppression_gather)
+
+layout non_max_suppression_gather_inst::calc_output_layout(non_max_suppression_gather_node const& node, kernel_impl_params const& impl_param) {
+    OPENVINO_THROW("Only calc_output_layouts should be used!");
+}
+
+template<typename ShapeType>
+std::vector<layout> non_max_suppression_gather_inst::calc_output_layouts(non_max_suppression_gather_node const& /*node*/,
+                                                                         const kernel_impl_params& impl_param) {
+    std::vector<layout> layouts;
+
+    auto desc = impl_param.typed_desc<non_max_suppression_gather>();
+    std::vector<ShapeType> output_shapes = { ShapeType{}, ShapeType{}, ShapeType{} };
+
+    auto& memory_deps = impl_param.memory_deps;
+    if (memory_deps.count(0)) {
```

Comment:
Can't third output (if present) with scalar value be used to get the actual_valid_num instead of iterating over the whole first buffer looking for -1?

Reply:
Third output(scalar) value just return entire input size. That's why I add NMS_Gather op to find actually valid output size which looking for first -1.

Reply:
@kelvinchoi-intel , are you sure? As I can see in the code, the variable `outputIdx` that is being saved into the 3d output is used for filling the 1st output data with -1 values you are reading:
https://github.com/openvinotoolkit/openvino/blob/182946beaaa44190a1cf5fe3d76c7dd1cf381805/src/plugins/intel_gpu/src/kernel_selector/cl_kernels/non_max_suppression_gpu_ref.cl#L878-L882
https://github.com/openvinotoolkit/openvino/blob/182946beaaa44190a1cf5fe3d76c7dd1cf381805/src/plugins/intel_gpu/src/kernel_selector/cl_kernels/non_max_suppression_gpu_ref.cl#L900

Am I missing something?

Reply:
>> valid_outputs - 1D tensor with 1 element of type output_type representing the total number of selected boxes.
Also seeing the spec doc, I agree with Sergey.

Reply:
Yes, Sergey is right. I was misunderstood about that.
I've updated to utilize 3rd output value.